xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager;
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.Rds;
+import io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.alts.GoogleDefaultChannelBuilder;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import io.grpc.xds.Bootstrapper.ChannelCreds;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+final class XdsClientImpl extends XdsClient {
+  private static final Logger logger = Logger.getLogger(XdsClientImpl.class.getName());
+
+  @VisibleForTesting
+  static final String ADS_TYPE_URL_LDS = "type.googleapis.com/envoy.api.v2.Listener";
+  @VisibleForTesting
+  static final String ADS_TYPE_URL_RDS =
+      "type.googleapis.com/envoy.api.v2.RouteConfiguration";
+
+  private final ManagedChannel channel;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timeService;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+  private final Stopwatch stopwatch;
+  // The node identifier to be included in xDS requests. Management server only requires the
+  // first request to carry the node identifier on a stream. It should be identical if present
+  // more than once.
+  private final Node node;
+
+  // Cached data for RDS responses, keyed by RouteConfiguration names.
+  // LDS responses indicate absence of RouteConfigurations and RDS responses indicate presence
+  // of RouteConfigurations.
+  // Optimization: only cache clusterName filed in the RouteConfiguration messages of RDS
+  // responses.
+  private final Map<String, String> routeConfigNamesToClusterNames = new HashMap<>();
+
+  @Nullable
+  private AdsStream adsStream;
+  @Nullable
+  private BackoffPolicy retryBackoffPolicy;
+  @Nullable
+  private ScheduledHandle rpcRetryTimer;
+
+  // Following fields are set only after the ConfigWatcher registered. Once set, they should
+  // never change.
+  @Nullable
+  private ConfigWatcher configWatcher;
+  // The host name portion of "xds:" URI that the gRPC client targets for.
+  @Nullable
+  private String hostName;
+  // The "xds:" URI (including port suffix if present) that the gRPC client targets for.
+  @Nullable
+  private String ldsResourceName;
+
+  XdsClientImpl(
+      // URI of the management server to be connected to.
```

Comment:
why not just have Javadoc comments to describe these params? Have a Javadoc for the constructor.

Reply:
This class is internal, we do not have convention to have formal Javadoc for internal classes (it's not necessary, they are internal implementations, not something exposed to users). You see a lot of head comments for methods, they are my personal convention, which are method specifications to help define what a method does (I usually write spec first before filling in implementation). Here some params are commented just to help readers understanding, they are good-to-have things. You definitely don't want to comment for params that already explain themselves.

Reply:
There's no need but since you already have the comment, why not make it Javadoc? That way the IDE shows the comment while browsing .